lms/static/scripts/frontend_apps/components/ComponentWithLoaderWrapper.tsx

```diff
@@ -14,40 +15,93 @@ export type LoaderOptions = {
   signal: AbortSignal;
 };
 
+type RouteModule = {
+  loader: (opts: LoaderOptions) => unknown;
+  default: FunctionalComponent;
+};
+
+const matchRoute = (parser: Parser, route: string, path: string) => {
+  const { pattern, keys } = parser(route);
+
+  // array destructuring loses keys, so this is done in two steps
+  const result = pattern.exec(path) || [];
+
+  // when parser is in "loose" mode, `$base` is equal to the
+  // first part of the route that matches the pattern
+  // (e.g. for pattern `/a/:b` and path `/a/1/2/3` the `$base` is `a/1`)
+  // we use this for route nesting
+  const [$base, ...matches] = result;
+
+  if ($base === undefined) {
+    return [false, null];
+  }
+
+  // an object with parameters matched, e.g. { foo: "bar" } for "/:foo"
+  // we "zip" two arrays here to construct the object
+  // ["foo"], ["bar"] → { foo: "bar" }
+  const groups = Object.fromEntries(keys.map((key, i) => [key, matches[i]]));
+
+  // convert the array to an instance of object
+  // this makes it easier to integrate with the existing param implementation
+  const obj = { ...matches };
+  // merge named capture groups with matches array
+  Object.assign(obj, groups);
+
+  return [true, obj];
+};
+
+function useRouteModule(
+  routeToModuleMap: Map<string, () => Promise<RouteModule>>,
+) {
+  const { parser } = useRouter();
+  const [location] = useLocation();
+
+  return useMemo(() => {
+    for (const [route, moduleResolver] of routeToModuleMap) {
+      const [matches, params] = matchRoute(parser, route, location);
+      if (matches) {
+        return { module: moduleResolver(), params: params ?? {} };
+      }
+    }
+
+    return undefined;
+  }, [location, parser, routeToModuleMap]);
+}
+
+const routesMap = new Map([
+  [
+    '/assignments/:assignmentId',
+    () => import('./dashboard/AssignmentActivity') as Promise<RouteModule>,
+  ],
+  [
+    '/courses/:courseId',
+    () => import('./dashboard/CourseActivity') as Promise<RouteModule>,
+  ],
+  [
+    '',
+    () => import('./dashboard/OrganizationActivity') as Promise<RouteModule>,
+  ],
+]);
```

Comment:
I hardcoded this here for simplicity, but the idea would be that this is passed as a prop by callers.